### ISSUE TITLE
Cannot make DALI work with tf.keras.

### ISSUE THREAD

**Issue by AlexZhurkevich** [NVIDIA/DALI#2623]
This is my DALI code:

```
tf.compat.v1.disable_eager_execution()

class SegmantationPipeline(Pipeline):

    def __init__(self, batch_size, device, device_id=0, num_threads=12, exec_pipelined=False):
        super(SegmantationPipeline, self).__init__(
            batch_size, num_threads, device_id)
        
        #CPU or GPU
        self.device = device

        self.reader = ops.TFRecordReader(
            #List of paths to TFRecord files.
            path='/media/Segmentation/DALITEST/512_train_R20-72 G02 2941-33_.tfrecord',
            #List of paths to index files. There should be one index file for every TFRecord file.
            index_path='/media/Segmentation/DALITEST/idx_files/train.idx',
            #A dictionary that maps names of the TFRecord features to extract to the feature type.
            features={
                'image/height': tfrec.FixedLenFeature([], tfrec.int64, 0),
                'image/width': tfrec.FixedLenFeature([], tfrec.int64, 0),
                'image/channels': tfrec.FixedLenFeature([], tfrec.int64, 0),
                'image/class/mask': tfrec.FixedLenFeature([], tfrec.string, ""),
                'image/image_filename': tfrec.FixedLenFeature([], tfrec.string, ""),
                'image/mask_filename': tfrec.FixedLenFeature([], tfrec.string, ""),
                'image/encoded': tfrec.FixedLenFeature([], tfrec.string, "")
            },
            #Determines whether to randomly shuffle data.
            random_shuffle=True,
            #Size of the buffer that is used for shuffling.
            initial_fill=2048,
            #Specifies the number of batches to be prefetched by the internal Loader.
            prefetch_queue_depth=5,
            #Determines whether the accessed data should be read ahead.
            read_ahead=True
        )

        #Decode image
        self.decode = ops.ImageDecoder(
            device='mixed' if device == 'gpu' else 'cpu',
            output_type=types.GRAY)
        
        #Normalize image
        self.norm = ops.Normalize(
            device=device,
            dtype=types.FLOAT)

    def define_graph(self):
        inputs = self.reader(name="Reader")
        images = self.decode(inputs["image/encoded"])
        masks = self.decode(inputs['image/class/mask'])

        if self.device == 'gpu':
            images = images.gpu()
            masks = masks.gpu()

        images = self.norm(images)
        masks = self.norm(masks)

        return (images, masks)

segmentation_pipeline = SegmantationPipeline(28, device='gpu', device_id=0)

```
Implementation is very close to the one with mnist example on the docs page, the only difference is that my input and label are two images, image and mask correspondingly. 
I can verify it is working by building it like this:
```
segmentation_pipeline.build()
pipe_out = segmentation_pipeline.run()
```
and running show_images like this:
```
#Sending masks
show_images(pipe_out[1].as_cpu())

#Show images code 
def show_images(image_batch):
    plt.style.use('grayscale')
    for j in range(28):
        img_chw = image_batch.at(j)
        plt.imshow(img_chw)
        plt.savefig("%d_.%s" % (j, "png"))


```
One of the images for example:
![20_](https://user-images.githubusercontent.com/36546909/105257420-bd3a6280-5b55-11eb-9590-0a243a82b540.png)

Seems like the pipeline is working, now I am getting rid of the .build() and .run() code and wrapping it into DALIDataset like this (close to what is on docs page):
```
shapes = (
    (28, 512, 512, 1),
    (28, 512, 512, 1))
dtypes = (
    tf.float32,
    tf.float32)


seg_set = dali_tf.DALIDataset(
    pipeline=segmentation_pipeline,
    batch_size=28,
    output_shapes=shapes,
    output_dtypes=dtypes,
    num_threads=12,
    device_id=0)
```
Also, at this step I can run a program and it gives me no errors so that's good, now let's connect it to my model in .fit method. I will also add model code just in case it might help:
```
def iou(y_true, y_pred):
    def f(y_true, y_pred):
        #print(y_pred)
        intersection = (y_true * y_pred).sum()
        union = y_true.sum() + y_pred.sum() - intersection
        x = (intersection + 1e-15) / (union + 1e-15)
        x = x.astype(np.float32)
        return x
    return tf.numpy_function(f, [y_true, y_pred], tf.float32)


def get_model(size, kernel_size):
    inputs = keras.Input((size, size, 1))

    ### [First half of the network: downsampling inputs] ###

    # Entry block
    x = layers.Conv2D(16, (kernel_size, kernel_size), strides=2, padding="same")(inputs)
    x = layers.BatchNormalization()(x)
    x = layers.Activation("relu")(x)

    previous_block_activation = x  # Set aside residual

    # Blocks 1, 2, 3 are identical apart from the feature depth.
    for filters in [32, 64, 128, 256]:
        x = layers.Activation("relu")(x)
        x = layers.SeparableConv2D(filters, (kernel_size, kernel_size), padding="same")(x)
        x = layers.BatchNormalization()(x)

        x = layers.Activation("relu")(x)
        x = layers.SeparableConv2D(filters, (kernel_size, kernel_size), padding="same")(x)
        x = layers.BatchNormalization()(x)

        x = layers.MaxPooling2D((kernel_size, kernel_size), strides=2, padding="same")(x)

        # Project residual
        residual = layers.Conv2D(filters, 1, strides=2, padding="same")(
            previous_block_activation
        )
        x = layers.add([x, residual])  # Add back residual
        previous_block_activation = x  # Set aside next residual

    ### [Second half of the network: upsampling inputs] ###

    for filters in [256, 128, 64, 32, 16]:
        x = layers.Activation("relu")(x)
        x = layers.Conv2DTranspose(filters, (kernel_size, kernel_size), padding="same")(x)
        x = layers.BatchNormalization()(x)

        x = layers.Activation("relu")(x)
        x = layers.Conv2DTranspose(filters, (kernel_size, kernel_size), padding="same")(x)
        x = layers.BatchNormalization()(x)

        x = layers.UpSampling2D(2)(x)

        # Project residual
        residual = layers.UpSampling2D(2)(previous_block_activation)
        residual = layers.Conv2D(filters, 1, padding="same")(residual)
        x = layers.add([x, residual])  # Add back residual
        previous_block_activation = x  # Set aside next residual

    # Add a per-pixel classification layer
    x = layers.Conv2D(1, 3, padding="same")(x)
    outputs = layers.Activation("sigmoid", dtype='float32', name='predictions')(x)

    # Define the model
    model = keras.Model(inputs, outputs)
    return model

metrics = ["acc", iou]
model = get_model(size, kernel_size)
model.compile(optimizer="adam", loss="binary_crossentropy", metrics=metrics)
model.summary()
model.fit(seg_set,
          epochs=EPOCHS,
          steps_per_epoch=3)
```
Not much different from mnist example, different models but linkage is the same. 
When I am launching it I am getting:
```
2021-01-20 19:46:33.384937: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1406] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 6206 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1070, pci bus id: 0000:01:00.0, compute capability: 6.1)
2021-01-20 19:46:33.444713: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:196] None of the MLIR optimization passes are enabled (registered 0 passes)
2021-01-20 19:46:33.522008: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 2199995000 Hz
WARNING:tensorflow:Expected a shuffled dataset but input dataset `x` is not shuffled. Please invoke `shuffle()` on input dataset.
DALI daliCreatePipeline( pipeline_handle, pipeline_def_.pipeline.c_str(), pipeline_def_.pipeline.length(), pipeline_def_.batch_size, pipeline_def_.num_threads, pipeline_def_.device_id, pipeline_def_.exec_separated, pipeline_def_.prefetch_queue_depth, pipeline_def_.cpu_prefetch_queue_depth, pipeline_def_.gpu_prefetch_queue_depth, pipeline_def_.enable_memory_stats) failed: Critical error when building pipeline:
Error when constructing operator: ImageDecoder, instance name: "__ImageDecoder_1", encountered:
std::bad_alloc
Current pipeline object is no longer valid.
2021-01-20 19:46:34.445879: F /tmp/pip-install-may5v6jm/nvidia-dali-tf-plugin-cuda110_d35391b46b0a49d4861d4194361ff907/dali_dataset_op.cc:144] Non-OK-status: InitPipeline(&pipeline_handle) status: Internal: DALI daliCreatePipeline( pipeline_handle, pipeline_def_.pipeline.c_str(), pipeline_def_.pipeline.length(), pipeline_def_.batch_size, pipeline_def_.num_threads, pipeline_def_.device_id, pipeline_def_.exec_separated, pipeline_def_.prefetch_queue_depth, pipeline_def_.cpu_prefetch_queue_depth, pipeline_def_.gpu_prefetch_queue_depth, pipeline_def_.enable_memory_stats) failed: Critical error when building pipeline:
Error when constructing operator: ImageDecoder, instance name: "__ImageDecoder_1", encountered:
std::bad_alloc
Current pipeline object is no longer valid.
Aborted (core dumped)
```
Seems like error is coming from ImageDecoder, not sure what is wrong with it, wasnt giving me any errors when I was testing it with show_images, any hints are very much appreciated!

I also have a couple additional questions. 
Lets say at a runtime I would like to unpack images and masks from TFRecords and augment them on the GPU on the fly, specifically per image and per mask get all 90 degree rotations as well as two flips of an original image, mask. How would I approach doing that in terms of structuring the data in define_graph method? Do augmentations and make a list of 6 images as well as a list of 6 masks and return two lists from define graph method? Or is there a better/recommended faster way of doing that, maybe I should just follow "Augmentation Gallery" doc?    
Your own opinion, do you think augmentations on the fly are worth it with DALI leveraging the speed of nvJPEG instead of saving pre-augmented images in TFRecords and just reading augmented images/masks from TFRecords at a runtime? 
Is disabling eager execution tf.compat.v1.disable_eager_execution() absolutely necessary when your are subclassing a DALI Pipeline?
Thank you so much for your time, I highly appreciate any help!

**Comment by JanuszL** [comment]
Hi,

TensorFlow allocates all available GPU memory in the greedy way and DALI has nothing left. Please try to limit TF memory usage, for example:
```
gpu_options = GPUOptions(per_process_gpu_memory_fraction=0.8)
config = ConfigProto(gpu_options=gpu_options)
```
like [in this example](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/frameworks/tensorflow/tensorflow-plugin-tutorial.html).

> How would I approach doing that in terms of structuring the data in define_graph method? Do augmentations and make a list of 6 images as well as a list of 6 masks and return two lists from define graph method?

As I understand, 6 is the batch size, so nothing special needs to be done.
You pipeline can look like:
```
        self.rotate_lin = ops.Rotate()
        # for masks use NN as we don't want any pixel interpolation
        self.rotate_nn= ops.Rotate(interp_type = types.DALIInterpType.INTERP_NN)
        (...)
        images = self.norm(images)
        masks = self.norm(masks)
        flip_vert = self.coin()
        images = self.flip(images , vertical = flip_vert)
        masks = self.flip(masks, vertical = flip_vert)
        angle = self.uniform()
        images = self.rotate_lin(images, angle = angle)
        masks = self.rotate_nn(masks , angle = angle)
```
`disable_eager_execution` is mostly to make the example compatible with TF1 and TF2.

**Comment by AlexZhurkevich** [comment]
Totally missed it, followed another doc that didnt have it. Thanks @JanuszL !
For people who will have the same problem and are using tensorflow 2+, enable eager execution and add 
```
gpus = tf.config.experimental.list_physical_devices('GPU')
if gpus:
    # Restrict TensorFlow to only allocate 5GB of memory on the first GPU
    try:
        tf.config.experimental.set_virtual_device_configuration(
            gpus[0],
            [tf.config.experimental.VirtualDeviceConfiguration(memory_limit=5120)])
        logical_gpus = tf.config.experimental.list_logical_devices('GPU')
        print(len(gpus), "Physical GPUs,", len(logical_gpus), "Logical GPUs")
    except RuntimeError as e:
        # Virtual devices must be set before GPUs have been initialized
        print(e)
```
Control mem usage with memory_limit, here is a [guide](https://www.tensorflow.org/guide/gpu#limiting_gpu_memory_growth).

Thanks for augmentations suggestions, Ill try it out!

**Comment by JanuszL** [comment]
I'm glad you finally managed to get it working.
